src/Symfony/Component/Form/FormFactory.php

```diff
@@ -126,7 +126,7 @@ public function createBuilderForProperty($class, $property, $data = null, array
 
         // user options may override guessed options
         if ($typeGuess) {
-            $options = array_merge($typeGuess->getOptions(), $options);
+            $options = array_replace_recursive($typeGuess->getOptions(), $options);
```

Comment:
Following the first assumption:
> But this may not be the case for all other options where the usage of some keys may be mutually exclusive.

Still we can pass `false` to the user option and it not will be displayed, i.e.:
```php
$form->add('foo', null, array('foo' => array('baz' => false, 'bar' => 'foobar')));

//...

$customGuessedOption = array('foo' => array('baz' => 'foobaz'));
```